### ISSUE TITLE
Necrotic upgrade/trait, balancing concerns

### ISSUE THREAD

**Issue by definitelynotrazu** [SlimeKnights/TinkersConstruct#4983]
### Minecraft Version

1.18.2

### Forge Version

40.1.73

### Mantle Version

1.9.27

### Tinkers' Construct Version

3.5.1.31

### Problem description

I've been experimenting with tinkers, by itself, in PvE and PvP with some friends of mine and I've noted some balancing concerns with the current necrotic mechanics. First I do want to preface this and mention the encyclopedia of tinkering references a percentage life steal and a percentage chance, it says "Heals 25% at level 1, 45% chance at level 2, 60% at level 3, 70% at level 4, and 75% at level 5" However from what I can tell from testing, there is no chance and instead simply a guaranteed % on crit. So the context of what I'm about to say, is built on the idea that this is the intended behavior and that the book entry is simply either a typo, or outdated. 

The first issue we all noted was that in combat, PvE and PvP but PvP most notably, the amount of damage you'd have to deal to kill a player vs the amount of damage they have to deal to kill you, the ratio of combat if you will, is heavily in favor of the necrotic user. For example, if two players had a broadaxe, any material assuming they both have the same, and one of them invests in necrotic, the opposing player has no upgrade that can be relative to the ratio of damage needed to beat the necrotic user other than necrotic. Sharpness will only increase +0.75 per level to a max of 5, swiftstrike will only improve speed by 5% per level, vs the grand 70% damage returned as health via necrotic at only level 4. This would force anyone in PvP to use necrotic in their weapon build and pushes necrotic to the meta in the majority of both PvP and PvE circumstances at least in vanilla and even in a number of packs. Rare arguments can be made for mobs that can one shot you, but I don't believe it's a common enough thing to be concerned about.

The second issue of note, is that necrotic is vastly superior on single target weapons, so far as to be underpowered with some AOE/crowd control weapons like a scythe. Purely due to the crit reliance of the mechanic and aoe weapons not being able to land crits on more than one enemy. But this is a lesser issue, still wanted to note it though. 

Also of last note, I did ask in the discord if I was missing any aspects of the mechanics of necrotic or if I was perchance facing some sort of bug and nobody mentioned anything so I do hope I haven't missed anything.

### Suggested solution

While I'm uncertain of a surefire solution, my friends and I discussed a potential idea that we did like. Shifting life steal away from instant heal on crit and into an over time effect that is created on kill. as currently to my knowledge we have no on kill perks in the arsenal. The idea is on kill, you're granted some level of healing over time based on the number of mobs you've killed within a time frame or within the time frame of the effect allowing you to stack it up (to a limit of course). Pushes it into a good crowd control option, that helps prevent the player from getting overwhelmed and keeps it in check for PvP. As for the exact numbers, I think the current numbers could work for the DOT effects. 25% at level 1, 45 at level 2, 70% at level 4, ect. As for the amount of time to distribute the effect over, I think 1 heart per second, per stack of the effect could work. 1 kill, 1 heart per second, 2 kills, 2 hearts per second, ect. Testing and tweaking would likely be needed but it seems reasonable imo, if a bit conservative for cautions sake. 

We discussed the idea of the same effect on hit rather than on kill, but we all agreed doing so would overshadow DOT effects in viability such as lacerating, fiery or decay from necronium and agreed on the kill perk. And we all thought the on kill perk would be a nice addition for weapons and play style building. So I hope you find the idea interesting at the very least.

### Alternatives considered

Returning damage dealt as hunger points, durability, saturation, stealing the enemies life literally as in actually increasing your health bar for a time instead of just healing your base health or even just turning down the numbers on necrotic to bring them closer in line with the other modifiers are all things I've seen discussed that I think would work fine. 

### Additional context

Apologies for the text wall, I had no short way to write it so thank you for reading it if you do.
Hope this isn't too long for a git post.

### Searched for existing enhancement?

Used the search bar, Checked the FAQ

**Comment by Insane96** [comment]
I like the idea, but would keep it on crit, as you can crit and run away to heal (instead of how it is right now as you can just crit and keep fighting). Making it on kill would make it not really useful in PvP as you would need to kill a player (and in a 1v1 becomes just useless).

By making it apply on crit would also make me have two weapons. One to deal tons of damage and then keep on going with another weapon while healing

**Comment by definitelynotrazu** [comment]
On crit would face the same issue I mentioned with on hit, where it'd be far superior on single target high damage weapons and rather weak on AOE/crowd control weapons and outshine the DOT modifiers in the effectiveness of combat, unless you brought down the numbers of healing over time to be in line with the DOT of lacerating and the likes per level. But other than that it could still work on crit, although in a 1v1 it would still be very useful if summons/pets are allowed. If summons/pets aren't allowed, then most weapons would be useless in a serious 1v1 other than the highest single target DPS which would be a bit boring for PvP if someone restricted it to that, at least imo. 
While I do still believe an on kill effect would be better, both for design and diversity, allowing you to push the numbers further without hurting balance, at the end of the day I could still back the idea of an on crit over time heal as long as the numbers were in line with the effectiveness of the DOT modifiers.

**Comment by KnightMiner** [comment]
Necrotic is impossible to balance. Anything we choose will either be too strong or too weak as minecraft combat is fundamentally flawed (it has been improving over the years, but this is still not a combat game)

In PvE, you need a large amount of healing for life steal to be worth it. In PvP, a large amount of life steal is too good of a defense as you noticed (though I'd argue it still does not outshine dual wielding max damage broad axes, as if you can kill your opponent in two quick hits life steal won't do them any good).

Applying life steal on kill is problematic as noted that would make it useless in PvP. Additionally, it would require a significant revamp as the current behavior heals based on the damage you deal, and the killing blow is rarely a significant portion of the enemy's health. I think that idea is better left to a different PvE focused modifier.

Applying the healing over time is a decent idea, could help balance out multiple hits by requiring you to wait for the full effect. Could also mean higher levels do not have to heal more but rather can heal faster. I do think the crit requirement is needed even in that case, the effect should not stack from multiple hits, so AOE would have no advantage. Not every modifier needs to be equally good on every tool, you are supposed to design a tool by considering good combinations.

**Comment by definitelynotrazu** [comment]
Yeah I agree, the combat system is a bit munted kek. Hopefully Daedelus' combat mod can get some traction, I think it's very promising.
I am intrigued however, in PvE do you refer to vanilla or modded? I ask since as far as I could test, even level 1 held it's value very well in PvE on broadaxes and the like. since you can still get into the 1 to 2 hit kill range on any non boss mob while having necrotic, it just becomes free healing with minor loss of TTK on bosses. Only time it fell off was during underwater fights and elytra combat, but those are niche cases.

As for healing over time on crit, but with the level increasing the speed rather than the amount, I like the idea. +1

**Comment by KnightMiner** [comment]
3.6.4.113 rebalances necrotic a bit, may or may not make more changes later.
